File: lib/CSharp/DataStructures/FenwickTree.cs
// Ref: https://github.com/camypaper/complib/blob/master/Library/Collections/Fenwick.cs
using System.Diagnostics;

// i, j は 全て 1-indexed です.
namespace Lib.FenwickTree
{
    public class FenwickTree
    {
        int n;
        long[] bit;
        int kMax = 1;
        public FenwickTree(int n)
        {
            this.n = n; bit = new long[n + 1];
            while ((kMax << 1) <= n) kMax <<= 1;
        }
        // Sum of [1, i].
        public long this[int i]
        {
            get
            {
                Debug.Assert(0 <= i && i <= n);
                if (i == 0) return 0;
                long sum = 0;
                for (; i > 0; i -= i & -i) sum += bit[i];
                return sum;
            }
        }
        public long this[int i, int j]
        {
            get
            {
                Debug.Assert(0 <= i && i <= j && j <= n);
                return i == 0 ? this[j] : this[j] - this[i - 1];
            }
        }
        public int LowerBound(long w)
        {
            if (w <= 0) return 0;
            int x = 0;
            for (int k = kMax; k > 0; k >>= 1)
            {
                if (x + k <= n && bit[x + k] < w)
                {
                    w -= bit[x + k];
                    x += k;
                }
            }
            return x + 1;
        }
        public void Add(int i, long x)
        {
            Debug.Assert(1 <= i && i <= n);
            for (; i < bit.Length; i += i & -i) bit[i] += x;
        }
        public long[] Items
        {
            get
            {
                var res = new long[n + 1];
                for (int i = 0; i < res.Length; i++) res[i] = this[i, i];
                return res;
            }
        }
    }
    public class RangeAddFenwickTree
    {
        int n;
        FenwickTree a, b;
        public RangeAddFenwickTree(int n)
        {
            this.n = n;
            a = new FenwickTree(n + 1);
            b = new FenwickTree(n + 1);
        }
        public void Add(int i, int j, long x)
        {
            Debug.Assert(1 <= i && i <= j && j <= n);
            a.Add(i, -(i - 1) * x);
            a.Add(j + 1, j * x);
            b.Add(i, x);
            b.Add(j + 1, -x);
        }
        // Sum of [1, i].
        public long this[int i]
        {
            get
            {
                Debug.Assert(0 <= i && i <= n);
                return i == 0 ? 0 : a[i] + b[i] * i;
            }
        }
        // Sum of [i, j].
        public long this[int i, int j]
        {
            get
            {
                Debug.Assert(0 <= i && i <= j && j <= n);
                return i == 0 ? this[j] : this[j] - this[i - 1];
            }
        }
        public long[] Items
        {
            get
            {
                var res = new long[n + 1];
                for (int i = 0; i < res.Length; i++) res[i] = this[i, i];
                return res;
            }
        }
    }
}

File: lib/rust/data_structures/fenwick_tree.rs
// Ref: https://github.com/rust-lang-ja/ac-library-rs/blob/master/src/fenwicktree.rs
pub struct FenwickTree {
    n: usize,
    bit: Vec<Vec<i64>>,
    uses_raq: bool,
}
impl FenwickTree {
    pub fn new(n: usize) -> Self {
        FenwickTree {
            n,
            bit: vec![vec![0; n]; 2],
            uses_raq: false,
        }
    }
    fn accum_sub(&self, is_raq: usize, mut i: usize) -> i64 {
        let mut sum = 0;
        while i > 0 {
            sum += self.bit[is_raq][i - 1];
            i &= i - 1;
        }
        sum
    }
    fn add_sub(&mut self, is_raq: usize, mut i: usize, x: i64) {
        while i <= self.n {
            self.bit[is_raq][i - 1] += x;
            i += i & i.wrapping_neg();
        }
    }
    /// performs data[i] += val;
    pub fn add(&mut self, mut i: usize, x: i64) {
        assert!(i <= self.n);
        i += 1;
        self.add_sub(0, i, x);
    }
    /// performs data[l], ..., data[r - 1] += val;
    pub fn range_add(&mut self, mut l: usize, mut r: usize, x: i64) {
        assert!(l <= r && r <= self.n);
        l += 1;
        r += 1;
        self.uses_raq = true;
        self.add_sub(0, l, -x * (l as i64 - 1));
        self.add_sub(0, r, x * (r as i64 - 1));
        self.add_sub(1, l, x);
        self.add_sub(1, r, -x);
    }
    fn accum(&self, i: usize) -> i64 {
        if self.uses_raq {
            self.accum_sub(0, i) + self.accum_sub(1, i) * i as i64
        } else {
            self.accum_sub(0, i)
        }
    }
    /// Returns data[l] + ... + data[r - 1].
    pub fn sum(&self, l: usize, r: usize) -> i64 {
        assert!(l <= r && r <= self.n);
        self.accum(r) - self.accum(l)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn fenwick_tree_works() {
        let mut ft = FenwickTree::new(5);
        // [1, 1, 1, 1, 1]
        ft.range_add(0, 5, 1);
        for i in 0..5 {
            assert_eq!(ft.sum(i, i + 1), 1);
        }

        let mut ft = FenwickTree::new(5);
        // [1, 2, 3, 4, 5]
        for i in 0..5 {
            ft.add(i, i as i64 + 1);
        }
        assert_eq!(ft.sum(0, 4), 10);
        assert_eq!(ft.sum(0, 5), 15);
        assert_eq!(ft.sum(3, 4), 4);
        assert_eq!(ft.sum(3, 5), 9);
    }
}
